### ISSUE TITLE
Change Request: Provide a way to disable all rules in config element

### ISSUE THREAD

**Issue by matwilko** [eslint/eslint#17360]
### ESLint version

8.44.0

### What problem do you want to solve?

While setting up a new project, I found a number of cases of different types of files that I wanted ESLint to process with only some very specific rules, and disable any other rules that might have been configured globally (no `files` prop) previously in the flat config array.

Pulling in many external plugin configurations meant I didn't have control over whether they scoped their rule configurations with `files`.

Usually this was for setting up linting on non-JS/TS files like JSON or YAML, but I also found a need for this when adding a custom rule to effectively ban JS files from the repo (in favour of TS only).

In these cases, I needed to turn off _every_ other rule that might have been enabled earlier in the array, both for performance reasons, but also so I didn't get lots of false-positive warnings from rules that wouldn't apply to those files.

Without built-in support for this, I had to write a post-processor of sorts, that took the "final" config array, scanned every element for every plugin, and every rule in those plugins, making a list of all possible rule names, and adding a `"ruleName": 0` entry to the rules object of the specific configs I had for those files. This is kinda janky, and kinda balloons the size of the config, but workable.

Given the existence of things like `eslint-plugin-json` and `eslint-plugin-yaml`, among presumably others, I think it's worth giving them a way to indicate in their config that all other rules need to be disabled for them.

### What do you think is the correct solution?

I think the solution I used in my codebase was workable as a general solution if ESLint were to take over processing internally.

I created a `const clearRules = Symbol("clearRules")` symbol that a config could import, and then add to their `rules` object, like so:

```
import { clearRules } from "eslint";

export default {
    files: ["**/*.xxx"],
    plugins: { ... },
    rules: {
        [clearRules]: true,
        "somePlugin/someRule": ["error"]
    }
};
```

The nice thing about using a symbol for this was that it doesn't interfere with either `config-array` or `eslint`, as it doesn't show up when enumerating keys/values/entries, and so it is basically ignored by both. It also cannot clash with any past or future rule name. The different syntax required for symbol usage also highlights to a user that they're doing something "special".

Note that I'm not proposing that `clearRules` would apply globally no matter where the config element is in the array - it would specifically only disable any rules that might have been enabled for the file _before_ the element, so users would need to be warned to place these sorts of configs at the end of their configuration arrays.

In order to actually process this through ESLint efficiently, I'd propose that the [`merge`](https://github.com/eslint/eslint/blob/b79b6fb64473969b426d086b484d2e29594a5e9a/lib/config/flat-config-schema.js#L341) method of the `rules` schema is updated such that: `const result = second[clearRules] ? { ...second } : { ...first, ...second };`.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by sam3k** [comment]
@matwilko you've said that you wrote a post-processor for this as a workaround. Do you mind sharing a link to it here?

**Comment by matwilko** [comment]
Unfortunately, it's company internal code that I can't immediately share without some management approvals :(

But, rewriting from scratch real quick as a rough outline:

```js
// clearRules.js
import { configs: eslintConfigs } from '@eslint/js';

function* getAllRuleNames(configs) {
    yield* Object.keys(eslintConfigs.all.rules);

    for (const config of configs) {
        for (const pluginName in config.plugins) {
            for (const ruleName in config.plugins[pluginName].rules) {
                yield ruleName;
            }
        }
    }
}

export const clearRules = Symbol('clearRules');

export function* disableAllRulesWhenClearRulesPresent(configs) {
    const allRulesDisabled = Object.fromEntries([...getAllRuleNames(configs)].map(name => [name, [0]]));

    for (const config in configs) {
        if (config.rules?.[clearRules]) {
            yield { ...config, rules: { ...allRulesDisabled, ...config.rules } };
        } else {
            yield config;
        }
    }
}
```

```ks
// postProcess.js
export default function postProcess(config, ...postProcessors) {
    return postProcessors.reduce((c, pp) => [...pp(c)], config);
}
```

```js
// eslint.config.js
import postProcess from './postProcess';
import { clearRules, disableAllRulesWhenClearRulesPresent } from './clearRules';
const configs = [
...,
...,
...,
{ files: ['**/*.xxx'], rules: { [clearRules]: true, "custom/ban-file": [2] },
...,
...
];

export default postProcess(configs, disableAllRulesWhenClearRulesPresent, ..., ...);
```

While this post-processing works, you should see the output of `eslint --print-config` for files it applies to when you have hundreds of rules - it's incomprehensible 😢

**Comment by nzakas** [comment]
We've had discussions about adding such a feature before, but it always ended up too complicated to move forward with. That said, if you'd like to put together an RFC, we can definitely evaluate it.

**Comment by matwilko** [comment]
I think flat config has probably made it much much simpler to reason about and implement, since it can just be part of the normal merge process 😄

I'm happy to put together a full proposal - are there any particular complications you recall off the top of your head, or links to previous discussions so I can address anything that's already come up? Are there any glaring holes in what's proposed above?

One thing I know I should ask about in advance: given your aversion to deleting rule configs in #17361, I assume the "solution" given above of simply dropping the `first` config is going to be unacceptable, and a slightly more complex solution of basically merging an "off" copy of `first` is going to be more palatable?

**Comment by nzakas** [comment]
@matwilko part of putting together the RFC is doing the research to find previous discussions. :) We have issues, PRs, and TSC meeting transcripts/notes that you can search through.

It's hard to evaluate a proposal in a vacuum. My advice for any RFC is to start with what you think is the right solution and plan on adapting it based on feedback. RFCs aren't a test you have to pass on the first try, they're a collaborative design process aimed at getting to the best solution.

**Comment by matwilko** [comment]
Sure, I was happy to do some digging, just asking in case there was anything at the top of your mind that might be relevant (especially stuff that I should probably avoid) or that might not appear in any prior art (if you've had plenty of time to think on it, there might be ideas in your head that weren't part of previous discussions 😄)

**Comment by Rec0iL99** [comment]
Not stale. Waiting for an RFC.

**Comment by nzakas** [comment]
Closing for now as there hasn't been any followup. If someone would like to create an RFC, we can reopen.
